[Docs] Add custom pipeline docs. (#1124)
open-mmlab/mmclassification: docs/en/advanced_guides/pipeline.md

# Customize Data Pipeline

## Design of Data pipelines

In the [new dataset tutorial](./datasets.md), we know that the dataset class use the `load_data_list` method
to initialize the entire dataset, and we save the information of every sample to a dict.

> Usually, to save memory usage, we won't load all images in the `load_data_list`, but only load when we use

Reviewer: ```suggestion
Usually, to save memory usage, we only load image paths and labels in the `load_data_list`, and load full image content when we use
```

## Modify the training/test pipeline

The data pipeline in MMClassification is pretty flexible. You can control almost every step of the data
preprocessing from the config file, but on the other hand, you may be confusing facing so many options.

Here is a common practice and guidance for image classification tasks.

### Loading

At the beginning of a data pipeline, we usually need to load image data from the file path.
[`LoadImageFromFile`](mmcv.transforms.LoadImageFromFile) is commonly used to do this task.

```python
train_pipeline = [
    dict(type='LoadImageFromFile'),
    ...
]
```

If you want to load data from files with special formats or special locations, you can [implement a new loading
transform](#add-new-data-transforms) and add it at the beginning of the data pipeline.

### Augmentation and other processing

During training, we usually need to do data augmentation to avoid overfitting. During the test, we also need to do
some data processing like resizing and cropping. These data transforms will be placed in the middle of the data
pipeline.

Here is a simple data augmentation recipe example. It will randomly resize and crop the input image to the

> specified scale, and randomly flip the image.

Reviewer: ```suggestion
specified scale, and randomly flip the image horizontally with probability.
```

[`RandomErasing`](mmcls.datasets.transforms.RandomErasing) as extra data augmentation method.

This configuration specified every detail of the data augmentation, and you can simply copy it to your own
config file to apply the data augmentations of the Swin-Transformer.

```python
bgr_mean = [103.53, 116.28, 123.675]
bgr_std = [57.375, 57.12, 58.395]

train_pipeline = [
    ...
    dict(type='RandomResizedCrop', scale=224, backend='pillow', interpolation='bicubic'),
    dict(type='RandomFlip', prob=0.5, direction='horizontal'),
    dict(
        type='RandAugment',
        policies='timm_increasing',
        num_policies=2,
        total_level=10,
        magnitude_level=9,
        magnitude_std=0.5,
        hparams=dict(
            pad_val=[round(x) for x in bgr_mean], interpolation='bicubic')),
    dict(
        type='RandomErasing',
        erase_prob=0.25,
        mode='rand',
        min_area_ratio=0.02,
        max_area_ratio=1 / 3,
        fill_color=bgr_mean,
        fill_std=bgr_std),
    ...
]
```

```{note}
Usually, the data augmentation part in the data pipeline handles only image-wise transforms, but not transforms

> like image normalization or mixup/cutmix. It's because we can do image normalization and mixup on batch data

Reviewer: ```suggestion
like image normalization or mixup/cutmix. It's because we can do image normalization and mixup/cutmix on batch data
```

to accelerate. To configure image normalization and mixup/cutmix, please use the [data preprocessor]
(mmcls.models.utils.data_preprocessor).
```

> ### formatting

Reviewer: Formatting
